libs/parallelism/algorithms/include/hpx/parallel/algorithms/unique.hpp

```diff
@@ -49,13 +502,26 @@ namespace hpx { namespace parallel { inline namespace v1 {
         /// \cond NOINTERNAL
 
         // sequential unique with projection function
-        template <typename FwdIter, typename Pred, typename Proj>
+        template <typename FwdIter, typename Sent, typename Pred, typename Proj>
         FwdIter sequential_unique(
-            FwdIter first, FwdIter last, Pred&& pred, Proj&& proj)
+            FwdIter first, Sent last, Pred&& pred, Proj&& proj)
         {
-            return std::unique(first, last,
-                util::compare_projected<Pred, Proj>(
-                    std::forward<Pred>(pred), std::forward<Proj>(proj)));
+            if (first == last)
+                return first;
+
+            util::compare_projected<Pred, Proj> pred_projected{
+                std::forward<Pred>(pred), std::forward<Proj>(proj)};
+
+            FwdIter result = first;
+            while (++first != last)
+            {
+                if (!HPX_INVOKE(pred_projected, *result, *first) &&
+                    ++result != first)
+                {
+                    *result = std::move(*first);
+                }
+            }
+            return ++result;
```

Comment:
If you implement the algorithm yourself, then there is no need to use `compare_projected` as you can directly invoke the projection. This might be beneficial for the compiler's optimization capabilities. In the simplest case this could be:
```
            while (++first != last)
            {
                if (!HPX_INVOKE(pred_projected, 
                        HPX_INVOKE(proj, *result), HPX_INVOKE(proj, *first)) &&
                    ++result != first)
                {
                    *result = std::move(*first);
                }
            }
```
More effort could improve the code even further by avoiding to invoke the projection more than once for a particular element.

Reply:
Thanks, made the changes